File: api/src/main/java/net/blueshell/api/mapper/ContributionPeriodMapper.java
package net.blueshell.api.mapper;


import net.blueshell.api.base.BaseMapper;
import net.blueshell.api.dto.ContributionPeriodDTO;
import net.blueshell.api.model.contribution.ContributionPeriod;
import org.mapstruct.*;

@Mapper(componentModel = "spring", nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
public abstract class ContributionPeriodMapper extends BaseMapper<ContributionPeriod, ContributionPeriodDTO> {

    @Mapping(target = "id")
    @Mapping(target = "startDate")
    @Mapping(target = "endDate")
    @Mapping(target = "halfYearFee")
    @Mapping(target = "fullYearFee")
    @Mapping(target = "alumniFee")
    @Mapping(target = "listId")
    @BeanMapping(ignoreByDefault = true)
    public abstract ContributionPeriodDTO toDTO(ContributionPeriod contributionPeriod);


    @Mapping(target = "id")
    @Mapping(target = "startDate")
    @Mapping(target = "endDate")
    @Mapping(target = "halfYearFee")
    @Mapping(target = "fullYearFee")
    @Mapping(target = "alumniFee")
    @BeanMapping(ignoreByDefault = true)
    public abstract ContributionPeriod fromDTO(ContributionPeriodDTO dto, @MappingTarget ContributionPeriod contributionPeriod);
}

File: api/src/main/kotlin/net/blueshell/api/contribution/web/mapper/ContributionPeriodMapper.kt
package net.blueshell.api.contribution.web.mapper

import net.blueshell.api.contribution.web.dto.ContributionPeriodDTO
import net.blueshell.api.shared.mapper.BaseMapper
import net.blueshell.api.contribution.persistence.ContributionPeriod
import org.springframework.stereotype.Component

@Component
class ContributionPeriodMapper : BaseMapper<ContributionPeriod, ContributionPeriodDTO>() {
    override fun fromDTO(dto: ContributionPeriodDTO): ContributionPeriod = fromDTO(dto, ContributionPeriod())

    fun fromDTO(dto: ContributionPeriodDTO, contributionPeriod: ContributionPeriod): ContributionPeriod {
        contributionPeriod.startDate = dto.startDate
        contributionPeriod.endDate = dto.endDate
        contributionPeriod.halfYearFee = dto.halfYearFee
        contributionPeriod.fullYearFee = dto.fullYearFee
        contributionPeriod.alumniFee = dto.alumniFee
        dto.version?.let { contributionPeriod.version = it }
        return contributionPeriod
    }

    override fun toDTO(contributionPeriod: ContributionPeriod): ContributionPeriodDTO {
        return ContributionPeriodDTO(
            startDate = contributionPeriod.startDate,
            endDate = contributionPeriod.endDate,
            halfYearFee = contributionPeriod.halfYearFee,
            fullYearFee = contributionPeriod.fullYearFee,
            alumniFee = contributionPeriod.alumniFee,
            listId = contributionPeriod.listId
        ).also { dto ->
            dto.id = contributionPeriod.id
            dto.version = contributionPeriod.version
        }
    }
}

fun ContributionPeriod.asDTO(mapper: ContributionPeriodMapper): ContributionPeriodDTO = mapper.toDTO(this)

fun ContributionPeriodDTO.asEntity(mapper: ContributionPeriodMapper): ContributionPeriod = mapper.fromDTO(this)
